server/controllers/categories.js

```diff
@@ -0,0 +1,141 @@
+const categoriesRouter = require('express').Router()
+const Category = require('../models/category')
+const Recipe = require('../models/recipe')
+
+
+categoriesRouter.get('/', async (request, response) => {
+  const user = request.user
+  if (!user) {
+    return response.status(401).json({ error: 'invalid user: no token provided' })
+  }
+
+  const categories = await Category
+    .find({userId : user.id}) // users can view only their own categories
+  response.json(categories)
+})
+
+
+categoriesRouter.post('/', async (request, response, next) => {
+  const { name } = request.body
+  const user = request.user
+
+  if (!user) {
+    return response.status(401).json({ error: 'invalid user: no token provided' })
+  }
+
+  const category = new Category({
+    name,
+    userId: user._id
+  })
+
+  // The format of the data is validated by Mongoose's validation
+  // functionality as specified in the recipe schema
+  try {
+    const savedCategory = await category.save()
+    response.status(201).json(savedCategory)
+  } catch (error) {
+    next(error) // give error to error handler middleware
+  }
+})
+
+
+categoriesRouter.delete('/:id', async (request, response, next) => {
```

Comment:
Deleting the "All" and other possible default categories should not be possible (as a note for this PR or for the future).